.github/workflows/daily-tests.yaml

```diff
@@ -8,404 +8,128 @@ on:
           ref: develop
-      - id: artifact-name
-        run: echo "name=$(date +"%Y-%m-%d_%H.%M.%S")-VEGA_X86" >> $GITHUB_OUTPUT
       - name: Build gem5
-        run: |
-          scons build/VEGA_X86/gem5.opt -j $(nproc)
+        run: scons build/${{ matrix.image }}/gem5.opt ${{ matrix.command-line }}
       - uses: actions/upload-artifact@v3
         with:
-          name: ${{ steps.artifact-name.outputs.name }}
-          path: build/VEGA_X86/gem5.opt
+          name: ${{ needs.name-artifacts.outputs.build-name }}${{ matrix.image }}
+          path: build/${{ matrix.image }}/gem5.opt
           retention-days: 5
       - run: echo "This job's status is ${{ job.status }}."
 
-
-  # This runs the unit tests for the build/ALL/unittests.debug build.
   unittests-all-debug:
+    strategy:
+      matrix:
+        type: [fast, debug]
     runs-on: [self-hosted, linux, x64, run]
     container: gcr.io/gem5-test/ubuntu-22.04_all-dependencies:latest
     timeout-minutes: 60
-
-    steps:
-      - uses: actions/checkout@v3
-        with:
-          # Scheduled workflows run on the default branch by default. We
-          # therefore need to explicitly checkout the develop branch.
-          ref: develop
-      - name: ALL/unittests.debug UnitTests
-        run: |
-          ls
-          scons build/ALL/unittests.debug -j $(nproc)
-
-  # This runs the unit tests for the build/ALL/unittests.fast build.
-  unittests-all-fast:
-    runs-on: [self-hosted, linux, x64, run]
-    container: gcr.io/gem5-test/ubuntu-22.04_all-dependencies:latest
-    timeout-minutes: 60
-
     steps:
       - uses: actions/checkout@v3
         with:
           # Scheduled workflows run on the default branch by default. We
           # therefore need to explicitly checkout the develop branch.
           ref: develop
-      - name: ALL/unittests.fast UnitTests
-        run: |
-          ls
-          scons build/ALL/unittests.fast -j $(nproc)
-
-# start running all of the long tests
-  testlib-long-arm-boot-tests:
+      - name: ALL/unittests.${{ matrix.type }} UnitTests
+        run: scons build/ALL/unittests.${{ matrix.type }} -j $(nproc)
+
+  # start running all of the long tests
+  testlib-long-tests:
+    strategy:
+      matrix:
+        test-type: [arm-boot-tests, fs, gem5_library_example_tests, gpu, insttest_se, learning_gem5, m5threads_test_atomic, memory, multi_isa, replacement-policies, riscv-boot-tests, stdlib, x86-boot-tests]
     runs-on: [self-hosted, linux, x64, run]
     container: gcr.io/gem5-test/ubuntu-22.04_all-dependencies:latest
-    needs: [build-gem5, build-gem5-ALL_CHI]
+    needs: [name-artifacts, build-gem5]
     timeout-minutes: 1440 # 24 hours
     steps:
     - uses: actions/checkout@v3
       with:
         # Scheduled workflows run on the default branch by default. We
         # therefore need to explicitly checkout the develop branch.
         ref: develop
+    # download all artifacts for each test
```

Comment:
Do you need to download all of them for all tests? If so, can you explain why in a comment?

Reply:
This one I'll explain here as well, but because we're now using a matrix, none of the long tests can start until all of the versions of gem5 within the matrix have been built.  I was toying around with the idea of still only downloading the necessary artifacts for each test, but the ways I was thinking of to do that would make this file a lot less readable than if I were to just download everything, plus it doesn't take significantly longer to wait for everything else to download.  There were also marketplace actions to download multiple artifacts in one step (which I think using would allow me to only download necessary artifacts but in a readable way), but they weren't from GitHub so I opted not to use those either